src/v/cloud_topics/reader/placeholder_extent.cc

```diff
@@ -0,0 +1,315 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
+ *
+ * Licensed as a Redpanda Enterprise file under the Redpanda Community
+ * License (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/redpanda-data/redpanda/blob/master/licenses/rcl.md
+ */
+
+#include "cloud_topics/reader/placeholder_extent.h"
+
+#include "cloud_io/basic_cache_service_api.h"
+#include "cloud_io/io_result.h"
+#include "utils/retry_chain_node.h"
+
+#include <seastar/core/lowres_clock.hh>
+
+#include <chrono>
+
+namespace experimental::cloud_topics {
+
+/// Map error codes from one type to another
+template<class Src, class Dst>
+struct errc_converter;
+
+template<>
+struct errc_converter<cloud_io::download_result, errc> {
```

Comment:
Don't like this mapping. Don't have better ideas but maybe you get some ideas by reading this tip https://www.boost.org/doc/libs/develop/libs/outcome/doc/html/tutorial/essential/conventions.html?

Reply:
this is not a public api, the whole thing lives inside .cc to begin with

Reply:
> Don't like this mapping

yeh i tend to agree. @nvartolomei can you say a bit more about what you don't like?

> this is not a public api, the whole thing lives inside .cc to begin with

i get what you are saying: it is contained, but i'm not sure it is relevant to nicoale's feedback about the pattern.

Reply:
In this case the error code enum from one module is mapped to error code enum from another. And the link says that this is a bad pattern and it's better to use `std::error_code` on api boundaries so you don't have to convert. The reasoning here is that `cloud_io::download_result` is not an `std::error_code`. Also, this is an exhaustive check. Sometimes the error code based solely on `enum class` is better because you can use `switch` stmt and rely on the compiler to check that the error check is exhaustive. In this case this utility function is not exposed outside and the stuff in .h returns `result` types which are based on `std::error_code`.

Reply:
Let's say I have a function `ss::future<result<iobuf>> download()` in the header. The `result` stores `std::error_code` which may contain different error categories. The module also has its own error code type. As a user of the module I'd still expect that all public functions / methods are returning error codes from the module itself, right? In this case I can still write an exhaustive error check.

Reply:
Also, Iceberg uses different approach with `boost::outcome` that returns error code which is not `std::error_code`. As Andrew explained to me the other day this was done to avoid boilerplate and to rely on `switch` statement during the error handling.

Reply:
thanks @Lazin this makes sense!